azure/Guidelines.md

```diff
@@ -0,0 +1,864 @@
+> 
+> :white_check_mark: **DO** establish a well-defined contract for the format of strings. For example, determine maximum length, legal characters, case-(in)sensitive comparisons, etc. Where possible, use standard formats, e.g. RFC3339 for date/time.
+> 
+> :white_check_mark: **DO** use strings formats that are well-known and easily parsable/formattable by many programming languages, e.g. RFC3339 for date/time.
+> 
+> :white_check_mark: **DO** ensure that information exchanged between your service and any client is "round-trippable" across multiple programming languages.
+> 
+> :white_check_mark: **DO** use [RFC3339] for date/time.
+> 
+> :white_check_mark: **DO** use [RFC4122] for UUIDs.
+
+> :heavy_check_mark: **YOU MAY** use JSON objects to group sub-fields together.
+> 
+> :heavy_check_mark: **YOU MAY** use JSON arrays if maintaining an order of values is required. Avoid arrays in other situations since arrays can be difficult and inefficient to work with, especially with ```JSON Merge Patch``` where the entire array needs to be read prior to any operation being applied to it.
+
+> :ballot_box_with_check: **YOU SHOULD** use JSON objects instead of arrays whenever possible.
+
+#### Enums & SDKs (Client libraries)
+It is common for strings to have an explicit set of values. These are often reflected in the OpenAPI definition as enumerations. These are extremely useful for developer tooling, e.g. code completion, and client library generation.
+
+However, it is not uncommon for the set of values to grow over the life of a service. For this reason, Microsoft's tooling uses the concept of an "extensible enum," which indicates that the set of values should be treated as only a *partial* list.
+This indicates to client libraries and customers that values of the enumeration field should be effectively treated as strings and that undocumented value may returned in the future. This enables the set of values to grow over time while ensuring stability in client libraries and customer code.
+
+> :ballot_box_with_check: **YOU SHOULD** use extensible enumerations unless you are positive that the symbol set will NEVER change over time.
+> 
+> :white_check_mark: **DO** document to customers that new values may appear in the future so that customers write their code today expecting these new values tomorrow.
+ 
+> :no_entry: **DO NOT** send "enum integers" over the wire.
+> 
+> :no_entry: **DO NOT** remove values from your enumeration list as this breaks customer code.
+
+#### Polymorphic types
+> :warning: **YOU SHOULD NOT** use polymorphic JSON types because they greatly complicate the customer code due to runtime dynamic casts and the introduction of new types in the future. 
+
+If you can't avoid them, then follow the guideline below.
+
+> :white_check_mark: **DO** define a ```kind``` field indicating the kind of the resource and include any kind-specific fields in the body. 
```

Comment:
Can we mandate that this filed must be first? Also, I am not sure this guideline is enough to be able to efficiently parse such JSON in .NET. I think it would be easier/faster if the kind property was outside of the object data, i.e. { kind : "...', value : ... }

Reply:
Including "kind" in this structure is the way that OpenAPI specifies how you do this and it will lead to problems. Because JSON is inherently unordered, and we can't guarantee how this will be serialized, we should not do anything that requires an order.

Reply:
This pattern is proper for JSON objects. It may be more difficult for .NET (and other languages) to parse but we also need to be true to REST/JSON customers.